conan/api/subapi/cache.py

```diff
@@ -181,13 +181,19 @@ def restore(self, path):
                 pkg_folder = pref_bundle["package_folder"]
                 out.info(f"Restore: {pref} in {pkg_folder}")
                 # We need to put the package in the final location in the cache
+                if os.path.exists(pkg_layout.package()):
+                    shutil.rmtree(pkg_layout.package())
```

Comment:
Point for discussion: I've made the change that whatever package is being restored will have priority over anything already in the cache. Should there be extra logging that this is happening, and should the local cache have priority?

Reply:
This sort of matches with what the tarfile.extractfile is doing. The extraction will overwrite any files already in the local cache with the new versions, but it will not remove any files that don't exist in the tar

Reply:
I think that in case of restoring a package, it shouldn't be additive, but the previous thing should be removed. The reason is that a package is a full-independent entity, with its checksums, its manifests listing contained files and checksums, etc. Merging 2 different sources can be problematic, it sounds that guaranteeing a clean restore of whatever is being restored sounds more appropiate. We could be discussing if the best is actually a no-op, as long as the hashes (recipe revision, package-id, package-revisions) are the same, in theory it shouldn't be necessary to remove and copy again, because the contents should be exactly the same by definition.

Reply:
Sounds reasonable to me, and good point about the checksums. I can have a look later to see how that could be used.

I think that the tar extraction should be changed to use a temporary directory, and then use the hashes to work out if there is something already in place that needs to be restored over. That would then allow proper no-op on a match, and prevent a potential dirty directory in the case that the package id is the same but the contents are different. Does that sound reasonable to you, and would you like me to work on that as part of this PR?

Reply:
>  but the contents are different

The contents cannot be different because they are always checksum-ed too into the recipe revision and the package revision. Each one gets always its own identifier both inside the "tgz" captured with ``conan cache save`` and in the Conan cache.

I am a bit concerned about this, as it might double the time depending on the system, unzipping in a different location, then moving it, I think this was one of the reasons why the untargz was done directly on the cache, taking advantage of the uniqueness of the folders as a function of revisions (checksums of contents)

Reply:
Ah gotcha, sorry I misunderstood what contributed to the package id.

Thanks for clarifying :D there won't be any change needed to the tar extraction